File: src/lion.js
/**
  A minimalist I18n (Internationalization) class with:
  - UI locale follows browser language setting, with EN fallback
  - Support secondary locale for multi-lingual users
  - standard JSON resource bundle with nested keys & placeholders
  - Auto-update translations using LLM during start-up
  - TM (Translation Memory) to store source/target pairs (TM records)
  - TM lookup to (1) reuse existing targets (2) detect source changes
 */

import { translateWithGPT } from './openai.js';
import { fetchJSON } from './utils.js';

const FALLBACK_LANGUAGE = 'en';
const LOCAL_TM = 'LOCAL_TM';

class I18n {
  constructor() {
    this.lang = getLanguageSetting();
    this.userLocale = this.lang.preferLocale;
    this.translations = {};
    this.TM = {};
  }

  async loadLocale(locale = FALLBACK_LANGUAGE) {
    try {
      const url = `/locales/${locale}.json`;
      const response = await fetchJSON(url);
      if (!response.ok) {
        console.warn(
          'Failed to load the',
          locale === FALLBACK_LANGUAGE
            ? 'required JSON resource file for the fallback language'
            : 'locale file',
          `'${locale}' from ${url}:`,
          response.statusText
        );
      } else {
        const data = await response.json();
        this.translations[locale] = data || {};
        return data;
      }
    } catch (error) {
      console.error(`Error loading locale '${locale}' :`, error);
    }
    return null;
  }

  t(key, params = {}, locale = this.userLocale) {
    const dict = this.translations[this.matchLocale(locale)] || {};
    const raw = I18n._dig(dict, key);
    if (typeof raw !== 'string') return key;
    return I18n._interpolate(raw, params);
  }

  matchLocale(locale) {
    if (!locale) return null;

    // Normalize: xx or xx-XX
    const norm = locale.replace('_', '-');
    const [lang, region] = norm.split('-');
    const full = region
      ? `${lang.toLowerCase()}-${region.toUpperCase()}`
      : lang.toLowerCase();

    // 1. Exact match (case-insensitive)
    const keys = Object.keys(this.translations);
    let found = keys.find((k) => k.toLowerCase() === full.toLowerCase());
    if (found) return found;

    // 2. Try language-only match
    found = keys.find((k) => k.toLowerCase().startsWith(lang.toLowerCase()));
    if (found) return found;

    // 3. Fallback
    if (keys.includes(FALLBACK_LANGUAGE)) return FALLBACK_LANGUAGE;

    return null;
  }

  static _dig(obj, path) {
    if (!obj || !path) return undefined;
    return path
      .split('.')
      .reduce(
        (acc, part) => (acc && acc[part] !== undefined ? acc[part] : undefined),
        obj
      );
  }

  static _interpolate(str, params) {
    return str.replace(/\{(\w+)\}/g, (_, name) => {
      return Object.prototype.hasOwnProperty.call(params, name)
        ? String(params[name])
        : `{${name}}`;
    });
  }

  // djb2 non-crypto hash function (source string key + value)
  static _hashKeyValue(key, value) {
    const str = `${key}\u0000${value}`;
    let hash = 5381;

    for (let i = 0; i < str.length; i += 1) {
      hash = (hash * 33 + str.charCodeAt(i)) >>> 0;
    }

    return hash.toString(16).padStart(8, '0');
  }

  // Flatten nested path to string leaves: returns [{ path, key, value }]
  static _flattenStrings(obj, basePath = '') {
    const out = [];
    if (!obj || typeof obj !== 'object') return out;
    for (const k of Object.keys(obj)) {
      const v = obj[k];
      const path = basePath ? `${basePath}.${k}` : k;
      if (typeof v === 'string') {
        out.push({ path, key: k, value: v });
      } else if (v && typeof v === 'object') {
        out.push(...I18n._flattenStrings(v, path));
      }
    }
    return out;
  }

  _setTMRecord(msgid, leafKey, locale, value) {
    if (!this.TM[msgid]) this.TM[msgid] = { key: leafKey };
    if (this.TM[msgid].key !== leafKey) {
      console.warn(
        `[TM] msgid key mismatch: keeping existing key='${this.TM[msgid].key}' new='${leafKey}'`
      );
    }
    this.TM[msgid][locale] = value;
  }

  /**
    Update Translation Memory (TM), keyed by a msgid hash (computed from source_json only)
    - TM structure:
        {
          <msgid>: {
            "key": "<leaf key>",
            {source_locale}: "<source value>",  # e.g. "en"
            {target_locale}: "<target value>"   # if available
          }
        }
    - Enforces unique leaf keys (ignoring paths). Warn on collisions and skip duplicates
    - If provided, target_json should have same set of keys. Warns for missing/extra keys.
   */
  updateTM(
    source_json,
    target_json = null,
    target_locale = null,
    source_locale = FALLBACK_LANGUAGE
  ) {
    if (!source_json || typeof source_json !== 'object')
      throw new Error('[TM] source_json error');
    if (target_json && !target_locale) {
      throw new Error(
        '[TM] updateTM: target_locale must be provided with target_json'
      );
    }

    const srcLeaves = I18n._flattenStrings(source_json);
    const srcPaths = new Set(srcLeaves.map((r) => r.path));
    let tgtLeaves = [];
    if (target_json) {
      tgtLeaves = I18n._flattenStrings(target_json);
      const tgtPaths = new Set(tgtLeaves.map((r) => r.path));
      // Compare and warn on key-path differences between source and target
      for (const p of srcPaths)
        if (!tgtPaths.has(p)) console.warn(`[TM] target missing path: ${p}`);
      for (const p of tgtPaths)
        if (!srcPaths.has(p))
          console.warn(`[TM] target has extra path not in source_json: ${p}`);
    }

    // Enforce unique leaf keys (ignore paths). Warns and skips duplicates.
    const seenLeafKeys = new Set();
    const srcUnique = [];
    for (const rec of srcLeaves) {
      if (seenLeafKeys.has(rec.key)) {
        console.warn(
          `[TM] Skipping duplicate leaf key '${rec.key}' on '${rec.path}'`
        );
        continue;
      }
      seenLeafKeys.add(rec.key);
      srcUnique.push(rec);
    }

    const tgtByPath = Object.create(null);
    for (const r of tgtLeaves) tgtByPath[r.path] = r;

    // Update TM using source msgid; attach target value if present for same leaf key
    for (const s of srcUnique) {
      const msgid = I18n._hashKeyValue(s.key, s.value);
      this._setTMRecord(msgid, s.key, source_locale, s.value);

      if (
        target_locale &&
        tgtByPath[s.path] &&
        typeof tgtByPath[s.path].value === 'string'
      ) {
        this._setTMRecord(msgid, s.key, target_locale, tgtByPath[s.path].value);
      }
    }

    return this.TM;
  }

  /**
    Return a translated version of source_json by leveraging what's already exist in TM.
    - target_json starts as a deep copy of source_json, to return in the same JSON structure
    - for each source string, compute msgid and look up target value under {target_locale}
    - What's ready to apply: If target found on TM, replace source value with target value
   */
  exportTM(target_locale, source_json = this.translations[FALLBACK_LANGUAGE]) {
    if (!source_json || typeof source_json !== 'object')
      throw new Error('[TM] source_json error');

    const target_json = structuredClone(source_json);
    const replaceIn = (obj) => {
      if (obj && typeof obj === 'object' && !Array.isArray(obj)) {
        for (const [k, v] of Object.entries(obj)) {
          if (typeof v === 'string') {
            const msgid = I18n._hashKeyValue(k, v);
            if (this.TM[msgid] && this.TM[msgid][target_locale]) {
              obj[k] = this.TM[msgid][target_locale];
            }
          } else if (v && typeof v === 'object' && !Array.isArray(v)) {
            replaceIn(v);
          }
        }
      }
    };
    replaceIn(target_json);
    return target_json;
  }

  /**
    Return missing translations after source changes, or return None if nothing is missing
    - What's need to translate: return source_json minus keys that already exist in TM
   */
  lookupTM(target_locale, source_json = this.translations[FALLBACK_LANGUAGE]) {
    if (!source_json || typeof source_json !== 'object')
      throw new Error('[TM] source_json error');

    const missing = structuredClone(source_json);
    const filterIn = (obj) => {
      if (obj && typeof obj === 'object' && !Array.isArray(obj)) {
        const toDelete = [];
        for (const [k, v] of Object.entries(obj)) {
          if (typeof v === 'string') {
            const msgid = I18n._hashKeyValue(k, v);
            if (this.TM[msgid] && this.TM[msgid][target_locale])
              toDelete.push(k);
          } else if (v && typeof v === 'object' && !Array.isArray(v)) {
            filterIn(v);
            if (!Object.keys(v).length) toDelete.push(k);
          }
        }
        for (const k of toDelete) delete obj[k];
      }
    };
    filterIn(missing);
    return Object.keys(missing).length ? missing : null;
  }

  async TranslateLocale(locale, load_target_file = false) {
    if (load_target_file) {
      const target_file = await i18n.loadLocale(locale);
      if (target_file)
        this.updateTM(
          this.translations[FALLBACK_LANGUAGE],
          target_file,
          locale
        );
    }
    const missing = this.lookupTM(locale);
    if (missing) {
      const translated = await translateWithGPT(
        JSON.stringify(missing),
        FALLBACK_LANGUAGE,
        locale
      );
      const urlParams = new URLSearchParams(window.location.search);
      if (urlParams.has('lion')) {
        console.debug('Translated:', JSON.stringify(translated, null, 2));
      }
      this.updateTM(missing, translated, locale);
    }
    const exported = this.exportTM(locale);
    this.translations[locale] = exported;
    return missing != null;
  }
}

export const i18n = new I18n();

export async function initi18n() {
  const urlParams = new URLSearchParams(window.location.search);
  if (urlParams.has('lion')) {
    console.debug('locale settings:', testI18n());
  }
  i18n.TM = JSON.parse(localStorage.getItem(LOCAL_TM)) || {};
  const source_file = await i18n.loadLocale();
  if (source_file) i18n.updateTM(source_file);

  const locale = urlParams.get('locale');
  if (isValidLocale(locale)) i18n.userLocale = locale;
}

export async function updateTranslation() {
  let updated = false;
  if (i18n.lang.preferLangCode !== FALLBACK_LANGUAGE) {
    updated = (await i18n.TranslateLocale(i18n.lang.preferLocale)) || updated;
  }

  if (
    i18n.lang.secondLangCode &&
    i18n.lang.secondLangCode !== FALLBACK_LANGUAGE
  ) {
    updated = (await i18n.TranslateLocale(i18n.lang.secondLocale)) || updated;
  }

  if (updated) {
    localStorage.setItem(LOCAL_TM, JSON.stringify(i18n.TM));
    console.debug('TM records:', i18n.TM);
  }
  console.debug('res-bundles: ', i18n.translations);
}

function isValidLocale(str) {
  // Matches:
  // - 2 letters (language code), e.g. "en"
  // - 2 letters + -/_ + 2 letters, e.g. "en-US" or "en_US"
  const regex = /^([a-zA-Z]{2}([-_][a-zA-Z]{2})?)$/;
  return regex.test(str);
}

export function testI18n() {
  const mock = new I18n();
  const mockSrc = {
    hello: 'Hello, {name}!',
    nested: { example: 'The answer is {value}.' },
  };
  const mockTgt = {
    hello: '你好，{name}',
    nested: { example: '答案是 {value}。' },
  };
  mock.translations[FALLBACK_LANGUAGE] = mockSrc;

  let tgt = mock.t('hello', { name: 'World' });
  console.debug('L10n test 1: ', tgt);
  console.assert(tgt === 'Hello, World!', 'Placeholder failed');

  mock.updateTM(mockSrc, mockTgt, 'zh');
  console.assert(
    mock.lookupTM('zh') === null,
    'lookupTM should return null when locale is ready'
  );
  //console.debug('TM records:', mock.TM);
  let missing = mock.lookupTM('fr');
  console.assert(
    missing && missing.hello === 'Hello, {name}!',
    'lookupTM should return missing in fallback locale'
  );
  let exported = mock.exportTM('zh');
  console.assert(
    exported.hello === '你好，{name}',
    'exportTM failed to return target string'
  );
  mock.translations['zh'] = exported;

  /*
  console.debug(
    'L10n test 2: ',
    (tgt = mock.t('nested.example', { value: 42 }, 'zh'))
  );
  console.assert(tgt === '答案是 42。', 'Nested failed');
  //console.debug('res-bundles: ', mock.translations);
  console.debug('L10n test 3: ', (tgt = mock.t('missing.key')));
  console.assert(tgt === 'missing.key', 'Missing key failed');

  const original_json = {
    app: {
      loading_text: 'Loading map…',
      caching_text: 'Finding landmarks…',
      search_placeholder: 'Search location…',
    },
    SettingDialog: {
      GOOGLE_MAPS_API_KEY: 'Google Maps API Key',
      OPENAI_API_KEY: 'OpenAI API Key',
      new_setting_prompt: 'Enter the name for the new setting:',
      duplicate_key_alert: 'This setting key already exists.',
    },
  };

  const translated_json = {
    app: {
      loading_text: '載入地圖…',
      caching_text: '搜尋地標…',
      search_placeholder: '搜尋此位置…',
    },
    SettingDialog: {
      GOOGLE_MAPS_API_KEY: 'Google 地圖 API 金鑰',
      OPENAI_API_KEY: 'OpenAI 金鑰',
      new_setting_prompt: '輸入新設定的名稱：',
      duplicate_key_alert: '此設定鍵已存在。',
    },
  };

  const modified_json = {
    app: {
      loading_text: 'Loading map, please wait…', // Changed value
      caching_text: 'Finding landmarks…', // Unchanged
      search_placeholder: 'Search for a place…', // Changed value
      new_feature_text: 'Welcome to the new map!', // New key/value
    },
    SettingDialog: {
      GOOGLE_MAPS_API_KEY: 'Google Maps API Key', // Unchanged
      OPENAI_API_KEY: 'AI API Key', // Changed value
      new_setting_prompt: 'Enter the name for the new setting:', // Unchanged
      duplicate_key_alert: 'Setting key already in use.', // Changed value
      extra_setting: 'New setting added.', // New key/value
    },
  };

  const partial_json = {
    app: { loading_text: 'Loading map' },
    SettingDialog: { GOOGLE_MAPS_API_KEY: 'Google Maps API Key' },
  };
  
  mock.updateTM(original_json, translated_json, 'zh');
  missing = mock.lookupTM('zh', modified_json);
  console.debug('Missing:', JSON.stringify(missing, null, 2));
  exported = mock.exportTM('zh', partial_json);
  console.debug('Exported:', JSON.stringify(exported, null, 2));
  */
  return mock.lang;
}

function getLanguageSetting() {
  const preferLocale = globalThis?.navigator?.language ?? FALLBACK_LANGUAGE;

  // Extract language-only code (e.g., 'en-US' -> 'en')
  const preferLangCode = preferLocale.split('-')[0].toLowerCase();

  // Preferred list of locales, for multi-lingual users
  const allLangs =
    globalThis && globalThis.navigator && navigator.languages?.length > 0
      ? navigator.languages
      : [preferLocale];

  // Find first locale with different language-only code
  let secondLocale =
    allLangs.find((locale) => {
      const langCode = locale.split('-')[0].toLowerCase();
      return langCode !== preferLangCode;
    }) ?? null;

  if (secondLocale === null) {
    secondLocale =
      preferLangCode !== FALLBACK_LANGUAGE ? FALLBACK_LANGUAGE : null;
  }

  const secondLangCode = secondLocale
    ? secondLocale.split('-')[0].toLowerCase()
    : null;

  return {
    preferLocale, // e.g., 'en-US'
    preferLangCode, // e.g., 'en'
    secondLocale, // e.g., 'fr-FR', 'en-US', or null
    secondLangCode,
  };
}

/**
 * Customize specific language code based on country
 * @param {string|Object} country - Country name string, or country object with name/code fields
 * @param {string} [code] - Country code (optional if country object has .code)
 * @returns {string} - custom locale code
 */
export function getCountryLanguage(country, code) {
  const { preferLangCode, secondLangCode } = i18n.lang;

  if (!country || (preferLangCode !== 'zh' && secondLangCode !== 'zh'))
    return i18n.userLocale;

  // Support country as string or object
  let countryName = '';
  let countryCode = '';
  if (typeof country === 'string') {
    countryName = country.toLowerCase();
    countryCode = (code || '').toUpperCase();
  } else {
    countryName = (country.name || '').toLowerCase();
    countryCode = (country.code || code || '').toUpperCase();
  }

  if (
    countryName.includes('hong kong') ||
    countryName.includes('香港') ||
    countryName.includes('macau') ||
    countryName.includes('澳門') ||
    countryName.includes('澳门') ||
    countryCode === 'HK'
  ) {
    return 'zh-HK';
  }

  if (
    countryName.includes('china') ||
    countryName.includes('中國') ||
    countryName.includes('中国') ||
    countryCode === 'CN'
  ) {
    return 'zh-CN';
  }

  if (
    countryName.includes('taiwan') ||
    countryName.includes('台灣') ||
    countryName.includes('台湾') ||
    countryCode === 'TW'
  ) {
    return 'zh-TW';
  }

  return i18n.userLocale;
}

export function setTooltip(element, key) {
  if (!element) return;
  element.setAttribute('data-i18n-title', key);
  element.title = i18n.t(key);
}

File: utils/lion.py
import copy
import hashlib
import json
import logging
import re
from pathlib import Path
from typing import Any, Dict, Optional

logger = logging.getLogger("lion")

FALLBACK_LANGUAGE = "en"
CUSTOM_SECOND_LOCALE = "zh-HK"
LOCAL_TM_FILE = Path("cache/local-tm.json")
LOCALES_DIR = Path("public/locales")
_LOCALE_REGEX = re.compile(r"^([a-zA-Z]{2}([-_][a-zA-Z]{2})?)$")


class I18n:
    def __init__(
        self,
        source_locale: str = FALLBACK_LANGUAGE,
        locales_dir: Path | str = LOCALES_DIR,
        tm_path: Path | str = LOCAL_TM_FILE,
        primary_locale: str = FALLBACK_LANGUAGE,
        second_locale: Optional[str] = CUSTOM_SECOND_LOCALE,
        user_locale: Optional[str] = None,
    ) -> None:
        self.source_locale = source_locale
        self.locales_dir = Path(locales_dir)
        self.tm_path = Path(tm_path)
        self.lang = {
            "preferLocale": primary_locale,
            "preferLangCode": _language_code(primary_locale),
            "secondLocale": second_locale,
            "secondLangCode": (
                _language_code(second_locale) if second_locale else None
            ),
        }
        self.user_locale = user_locale or primary_locale
        self.translations: Dict[str, Dict[str, Any]] = {}
        self.TM: Dict[str, Dict[str, Any]] = {}

    def load_locale(self, locale: str = FALLBACK_LANGUAGE) -> Optional[Dict[str, Any]]:
        locale_path = self.locales_dir / f"{locale}.json"
        if not locale_path.exists():
            logger.debug("Locale resource not found: %s", locale_path)
            return None

        try:
            data = json.loads(locale_path.read_text(encoding="utf-8"))
        except Exception as exc:
            logger.error("Error loading locale '%s': %s", locale, exc)
            return None

        if not isinstance(data, dict):
            logger.error("Locale '%s' must be a JSON object", locale)
            return None

        self.translations[locale] = data
        return data

    def load_tm(self) -> Dict[str, Dict[str, Any]]:
        if not self.tm_path.exists():
            self.TM = {}
            return self.TM

        try:
            data = json.loads(self.tm_path.read_text(encoding="utf-8"))
        except Exception as exc:
            logger.error("Error loading TM '%s': %s", self.tm_path, exc)
            self.TM = {}
            return self.TM

        self.TM = data if isinstance(data, dict) else {}
        return self.TM

    def save_tm(self) -> None:
        self.tm_path.parent.mkdir(parents=True, exist_ok=True)
        self.tm_path.write_text(
            json.dumps(self.TM, indent=2, ensure_ascii=False) + "\n",
            encoding="utf-8",
        )

    def t(
        self,
        key: str,
        params: Optional[Dict[str, Any]] = None,
        locale: Optional[str] = None,
    ) -> str:
        params = params or {}
        matched_locale = self.match_locale(locale or self.user_locale)
        if not matched_locale:
            return key

        raw = self._dig(self.translations.get(matched_locale, {}), key)
        if not isinstance(raw, str):
            return key
        return self._interpolate(raw, params)

    def match_locale(self, locale: Optional[str]) -> Optional[str]:
        if not locale:
            return (
                self.source_locale
                if self.source_locale in self.translations
                else None
            )

        norm = locale.replace("_", "-")
        parts = norm.split("-")
        lang = parts[0].lower()
        full = lang if len(parts) == 1 else f"{lang}-{parts[1].upper()}"

        keys = list(self.translations.keys())
        found = next((k for k in keys if k.lower() == full.lower()), None)
        if found:
            return found

        found = next((k for k in keys if k.lower().startswith(lang)), None)
        if found:
            return found

        return self.source_locale if self.source_locale in self.translations else None

    def update_tm(
        self,
        source_json: Dict[str, Any],
        target_json: Optional[Dict[str, Any]] = None,
        target_locale: Optional[str] = None,
        source_locale: str = FALLBACK_LANGUAGE,
    ) -> Dict[str, Dict[str, Any]]:
        if not isinstance(source_json, dict):
            raise ValueError("[TM] source_json error")
        if target_json is not None and not target_locale:
            raise ValueError(
                "[TM] update_tm: target_locale must be provided with target_json"
            )

        src_leaves = self._flatten_strings(source_json)
        src_paths = {record["path"] for record in src_leaves}
        tgt_leaves: list[Dict[str, str]] = []

        if target_json is not None:
            if not isinstance(target_json, dict):
                raise ValueError("[TM] target_json error")
            tgt_leaves = self._flatten_strings(target_json)
            tgt_paths = {record["path"] for record in tgt_leaves}
            for path in sorted(src_paths - tgt_paths):
                logger.warning("[TM] target missing path: %s", path)
            for path in sorted(tgt_paths - src_paths):
                logger.warning(
                    "[TM] target has extra path not in source_json: %s", path
                )

        seen_leaf_keys = set()
        src_unique = []
        for record in src_leaves:
            if record["key"] in seen_leaf_keys:
                logger.warning(
                    "[TM] Skipping duplicate leaf key '%s' on '%s'",
                    record["key"],
                    record["path"],
                )
                continue
            seen_leaf_keys.add(record["key"])
            src_unique.append(record)

        tgt_by_path = {record["path"]: record for record in tgt_leaves}
        for record in src_unique:
            msgid = self._hash_key_value(record["key"], record["value"])
            self._set_tm_record(msgid, record["key"], source_locale, record["value"])

            target_record = tgt_by_path.get(record["path"])
            if (
                target_locale
                and target_record
                and isinstance(target_record.get("value"), str)
            ):
                self._set_tm_record(
                    msgid,
                    record["key"],
                    target_locale,
                    target_record["value"],
                )

        return self.TM

    def export_tm(
        self,
        target_locale: str,
        source_json: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        source_json = source_json or self.translations.get(self.source_locale)
        if not isinstance(source_json, dict):
            raise ValueError("[TM] source_json error")

        target_json = copy.deepcopy(source_json)

        def replace_in(obj: Dict[str, Any]) -> None:
            for key, value in obj.items():
                if isinstance(value, str):
                    msgid = self._hash_key_value(key, value)
                    record = self.TM.get(msgid, {})
                    if isinstance(record.get(target_locale), str):
                        obj[key] = record[target_locale]
                elif isinstance(value, dict):
                    replace_in(value)

        replace_in(target_json)
        return target_json

    def lookup_tm(
        self,
        target_locale: str,
        source_json: Optional[Dict[str, Any]] = None,
    ) -> Optional[Dict[str, Any]]:
        source_json = source_json or self.translations.get(self.source_locale)
        if not isinstance(source_json, dict):
            raise ValueError("[TM] source_json error")

        missing = copy.deepcopy(source_json)

        def filter_in(obj: Dict[str, Any]) -> None:
            keys_to_delete = []
            for key, value in obj.items():
                if isinstance(value, str):
                    msgid = self._hash_key_value(key, value)
                    if isinstance(self.TM.get(msgid, {}).get(target_locale), str):
                        keys_to_delete.append(key)
                elif isinstance(value, dict):
                    filter_in(value)
                    if not value:
                        keys_to_delete.append(key)
            for key in keys_to_delete:
                del obj[key]

        filter_in(missing)
        return missing if missing else None

    def translate_locale(
        self, locale: str, load_target_file: bool = False
    ) -> bool:
        if load_target_file:
            target_file = self.load_locale(locale)
            if target_file:
                self.update_tm(
                    self.translations[self.source_locale],
                    target_file,
                    locale,
                    source_locale=self.source_locale,
                )

        missing = self.lookup_tm(locale)
        updated = False
        if missing:
            translated = self._translate_missing(missing, locale)
            if translated:
                self.update_tm(
                    missing,
                    translated,
                    locale,
                    source_locale=self.source_locale,
                )
                updated = True

        self.translations[locale] = self.export_tm(locale)
        return updated

    def _translate_missing(
        self, missing: Dict[str, Any], target_locale: str
    ) -> Optional[Dict[str, Any]]:
        try:
            from services.llm_service import LLMService

            return LLMService.translate_json_resource(
                missing, self.source_locale, target_locale
            )
        except Exception as exc:
            logger.error("Error translating locale '%s': %s", target_locale, exc)
            return None

    def _set_tm_record(
        self, msgid: str, leaf_key: str, locale: str, value: str
    ) -> None:
        if msgid not in self.TM:
            self.TM[msgid] = {"key": leaf_key}
        if self.TM[msgid].get("key") != leaf_key:
            logger.warning(
                "[TM] msgid key mismatch: keeping existing key='%s' new='%s'",
                self.TM[msgid].get("key"),
                leaf_key,
            )
        self.TM[msgid][locale] = value

    @staticmethod
    def _dig(obj: Dict[str, Any], path: str) -> Any:
        current: Any = obj
        for part in path.split("."):
            if not isinstance(current, dict) or part not in current:
                return None
            current = current[part]
        return current

    @staticmethod
    def _interpolate(text: str, params: Dict[str, Any]) -> str:
        return re.sub(
            r"\{(\w+)\}",
            lambda match: str(params.get(match.group(1), match.group(0))),
            text,
        )

    @staticmethod
    def _hash_key_value(key: str, value: str) -> str:
        data = f"{key}\u0000{value}".encode("utf-8")
        return hashlib.md5(data).hexdigest()[:8]

    @staticmethod
    def _flatten_strings(
        obj: Dict[str, Any], base_path: str = ""
    ) -> list[Dict[str, str]]:
        out: list[Dict[str, str]] = []
        for key, value in obj.items():
            path = f"{base_path}.{key}" if base_path else key
            if isinstance(value, str):
                out.append({"path": path, "key": key, "value": value})
            elif isinstance(value, dict):
                out.extend(I18n._flatten_strings(value, path))
        return out


def init_i18n(
    user_locale: Optional[str] = None,
    second_locale: Optional[str] = CUSTOM_SECOND_LOCALE,
    load_target_file: bool = False,
    locales_dir: Path | str = LOCALES_DIR,
    tm_path: Path | str = LOCAL_TM_FILE,
) -> I18n:
    i18n = I18n(
        source_locale=FALLBACK_LANGUAGE,
        locales_dir=locales_dir,
        tm_path=tm_path,
        primary_locale=FALLBACK_LANGUAGE,
        second_locale=second_locale,
        user_locale=user_locale or FALLBACK_LANGUAGE,
    )

    i18n.load_tm()
    source_file = i18n.load_locale(FALLBACK_LANGUAGE)
    if source_file:
        i18n.update_tm(source_file, source_locale=FALLBACK_LANGUAGE)

    active_locale = (
        i18n.user_locale
        if is_valid_locale(i18n.user_locale)
        else FALLBACK_LANGUAGE
    )
    if active_locale != FALLBACK_LANGUAGE and source_file:
        updated = i18n.translate_locale(
            active_locale, load_target_file=load_target_file
        )
        if updated:
            i18n.save_tm()

    return i18n


def is_valid_locale(locale: Optional[str]) -> bool:
    return bool(locale and _LOCALE_REGEX.fullmatch(locale))


def _language_code(locale: Optional[str]) -> Optional[str]:
    if not locale:
        return None
    return locale.replace("_", "-").split("-")[0].lower()


def get_country_language(country, code=None, locale="en"):
    """
    Customize specific language code based on country.
    Accepts country as a string (country name) or a dict (with 'name' and/or 'code').
    """
    if not country or _language_code(locale) != "zh":
        return locale

    country_name = ""
    country_code = ""
    if isinstance(country, str):
        country_name = country.lower()
        country_code = (code or "").upper()
    else:
        country_name = (country.get("name", "") or "").lower()
        country_code = (country.get("code", "") or code or "").upper()

    if (
        "hong kong" in country_name
        or "香港" in country_name
        or "macau" in country_name
        or "澳門" in country_name
        or "澳门" in country_name
        or country_code == "HK"
    ):
        return "zh-HK"

    if (
        "china" in country_name
        or "中國" in country_name
        or "中国" in country_name
        or country_code == "CN"
    ):
        return "zh-CN"

    if (
        any(name in country_name for name in ["taiwan", "台灣", "台湾"])
        or country_code == "TW"
    ):
        return "zh-TW"

    return locale
